src/view/segmentation.js

```diff
@@ -0,0 +1,248 @@
+/* globals newspack_popups_view */
+
+// Ensure we have a newspackRAS array.
+window.newspackRAS = window.newspackRAS || [];
+
+// Cached instance of RAS reader data object.
+let readerData;
+
+/**
+ * Log debugging data if WP_DEBUG is set.
+ *
+ * @param {string} key  Key name for debug data.
+ * @param {any}    data Data to log.
+ */
+const debug = ( key, data ) => {
+	if ( ! newspack_popups_view.debug ) {
+		return;
+	}
+
+	window.newspack_popups_debug = window.newspack_popups_debug || {};
+	window.newspack_popups_debug[ key ] = data;
+};
+
+/**
+ * Log the current pageview as a visit event in reader activity data.
+ */
+const logVisit = () => {
+	const { visit } = newspack_popups_view;
+
+	if ( ! visit.post_id || ! visit.post_type ) {
+		return;
+	}
+
+	// If the visit is an article, increment the lifetime view count.
+	if ( 'post' === visit.post_type ) {
+		const articleViewCount = readerData.store.get( 'articleViewCount' ) || 0;
+		readerData.store.set( 'articleViewCount', articleViewCount + 1 );
+	}
+
+	// If the visit has categories.
+	if ( visit.categories ) {
+		visit.categories = visit.categories.split( ',' ).map( id => parseInt( id ) );
+
+		// Add categories to lifetime category view count.
+		const categoryViewCounts = readerData.store.get( 'categoryViewCounts' ) || {};
+		visit.categories.forEach( categoryId => {
+			if ( ! categoryViewCounts[ 'cat' + categoryId ] ) {
+				categoryViewCounts[ 'cat' + categoryId ] = 0;
+			}
+			categoryViewCounts[ 'cat' + categoryId ]++;
+		} );
+		readerData.store.set( 'categoryViewCounts', categoryViewCounts );
+	}
+
+	window.newspackRAS.push( [ 'view', visit ] );
+};
+
+/**
+ * Get all prompts on the page.
+ *
+ * @return {Array} Array of prompt elements.
+ */
+const getPrompts = () => {
+	return [ ...document.querySelectorAll( '.newspack-popup' ) ];
+};
+
+/**
+ * Does the current reader match the given segment?
+ *
+ * @param {Object} segment Segment configuration object.
+ * @return {boolean} True if the reader matches the segment, false if not.
+ */
+const doesReaderMatchSegment = segment => {
+	const { configuration = {} } = segment;
+	let matches = true;
+
+	// Article view count.
+	const articleViewCount = readerData.store.get( 'articleViewCount' ) || 0;
+	if (
+		configuration.min_posts &&
+		configuration.min_posts > 0 &&
+		articleViewCount < configuration.min_posts
+	) {
+		matches = false;
+	}
+	if (
+		configuration.max_posts &&
+		configuration.max_posts > 0 &&
+		articleViewCount > configuration.max_posts
+	) {
+		matches = false;
+	}
```

Comment:
I know it's a draft but I wanted to register that we might wish to use a more abstract logic for implementing the matching logic for segmentation criteria.

In the spirit of having criteria as an extensible aspect of the plugin, we can think of them as something you configure and register, along with its matching logic:

```js
// Article view count criteria registration
const criteria = {
	name: 'Article view count',
	// ... other necessary metadata
	matchCallback: () => {
		const articleViewCount = readerData.store.get( 'articleViewCount' ) || 0;
		return true;
	}
}
registerCriteria( 'articleViewCount', criteria );
```

The `registerCriteria()` function would store a map of criteria. The `doesReaderMatchSegment()` function would read from that map to filter the relevant criteria for the given segment and execute each criteria callback.

Reply:
The criteria could even be the one creating any necessary data, either from an `init` callback in its registration schema or as part of its overall logic, imagining it as something located in a `criteria/articleViewCount.js` file. 🤔

Reply:
In addition to that, in the same discussion but for a separate PR, we should also consider a criteria registration for the backend in the same fashion

Reply:
This is a great idea. I think I grasp what you're suggesting and will give it a go, but I might huddle with you if I run into any blockers since you seem to have a strong idea of how we could set this up.